### PR TITLE
Support array of preferred ports

### LINKED ISSUE #19: Array of preferred ports instead of just one number
I think it could be interesting to have the ability to pass an array of preferred ports instead of just one before get-port gives up and tries a random port.

Would this be possible?

### PR DESCRIPTION
**Pull request by nagen010**
I have added new key ports to options that accept a list of preferred port numbers. Updated tests and documentation. It closes #19. Please review and let me know if I need to update something. Thanks.

### CODE REVIEW

**Comment by nagen010** [comment]
@sindresorhus Please review and let me know if it is okay :)

File: readme.md [hunk 1]
```diff
@@ -47,6 +55,12 @@ Type: `number`
 
 The preferred port to use.
 
+##### ports
```
[the remark this answers is not in the document]
Comment (nagen010) [reply readme.md]: Sure. I will just use the port option to handle both inputs

File: index.js [hunk 2]
```diff
@@ -1,6 +1,22 @@
 'use strict';
 const net = require('net');
 
+const isAvailable = (options, callback) => {
```
[the remark this answers is not in the document]
Comment (nagen010) [reply index.js]: Sure. I will keep the promise.

**Comment by nagen010** [comment]
@sindresorhus xo latest version v0.21.0 dropped support for Node.js 4 https://github.com/xojs/xo/releases/tag/v0.21.0. Shall I drop Node.js 4 support?

File: index.js [hunk 3]
```diff
@@ -10,19 +22,43 @@ const getPort = options => new Promise((resolve, reject) => {
...
+					})
+					.catch(() => {
+						return new Promise((resolve, reject) => reject());
+					});
+			});
+		}, Promise.reject())
+			.then(port => resolve(port))
+			.catch(() => reject());
+	} else {
+		isAvailable(options)
```
[the remark this answers is not in the document]
Comment (nagen010) [reply index.js]: Removed if-else block.

File: index.js [hunk 4]
```diff
@@ -1,6 +1,18 @@
...
+	server.on('error', reject);
+	server.listen(options, () => {
+		const {port} = server.address();
+		server.close(() => {
+			resolve(port);
+		});
+	});
+});
+
 const getPort = options => new Promise((resolve, reject) => {
```
[the remark this answers is not in the document]
Comment (nagen010) [reply index.js]: @sindresorhus It looks like util.promisify function was included only in NodeJS version 8.

[the remark this answers is not in the document]
Comment (sindresorhus) [reply index.js]: Use https://github.com/sindresorhus/pify

**Comment by nagen010** [comment]
@evilebottnawi if I understand correctly, I think there is no dependency to webpack-serve right?

**Comment by evilebottnawi** [comment]
@nagen010 https://github.com/webpack-contrib/webpack-serve/blob/master/package.json#L39

File: index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (nagen010) [reply index.js]: // @sindresorhus sorry. I am not sure how to use pify here. Could you please help me with some examples? Thanks.

File: index.js [hunk 5]
```diff
@@ -1,28 +1,60 @@
...
+				.catch(() => {
+					return new Promise((resolve, reject) => reject());
+				});
+		});
+	}, Promise.reject())
+		.then(port => resolve(port))
+		.catch(() => reject());
+});
+
+function getOptions(portNumber, hostname) {
```
Comment (acostalima) [inline index.js]: Or simply:

```
function getOptions(port, host) {
   return { port, host };
}
```

which makes me wonder if this function is really that necessary.

File: index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (acostalima) [reply index.js]: The follow should do:

```
const getPort = pify((options) => { ... });
```

You will have to adapt the `resolve` and `reject` logic accordingly.

File: index.js [hunk 6]
```diff
@@ -1,28 +1,60 @@
...
+			return isAvailable(input)
+				.then(port => {
+					return port;
+				})
+				.catch(() => {
+					return new Promise((resolve, reject) => reject());
+				});
+		});
+	}, Promise.reject())
+		.then(port => resolve(port))
```
[the remark this answers is not in the document]
Comment (nagen010) [reply index.js]: Thanks for the feedback!

File: index.js [hunk 7]
```diff
@@ -1,28 +1,43 @@
...
+		return seq.catch(() => {
+			const input = getOptions(port, options.host);
+			return isAvailable(input)
+				.then(port => port)
+				.catch(Promise.reject.bind(Promise));
+		});
+	}, Promise.reject()).then(resolve).catch(reject);
+});
+
+function getOptions(port, host) {
```
Comment (acostalima) [inline index.js]: Don't you think this function can be removed after all?

The following will do just fine:

```
if (typeof options === 'number') {
   options = { port: options };
}
```

ànd

```
isAvailable(Object.assign({}, options, { port }));
```

With `Object.assign`, you retain the previously implemented behavior which is the ability to eventually pass more options to `server.listen`.

File: index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (acostalima) [reply index.js]: @nagen010 what have you tried so far with `pify`?

**Pull request merged**